// @dharmaprotocol/contracts/dist/artifacts/ts/ERC721Basic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC721Basic = {
    "contractName": "ERC721Basic",
    "abi": [
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": true,
                    "name": "_from",
                    "type": "address"
                },
                {
                    "indexed": true,
                    "name": "_to",
                    "type": "address"
                },
                {
                    "indexed": false,
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "Transfer",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": true,
                    "name": "_owner",
                    "type": "address"
                },
                {
                    "indexed": true,
                    "name": "_approved",
                    "type": "address"
                },
                {
                    "indexed": false,
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "Approval",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": true,
                    "name": "_owner",
                    "type": "address"
                },
                {
                    "indexed": true,
                    "name": "_operator",
                    "type": "address"
                },
                {
                    "indexed": false,
                    "name": "_approved",
                    "type": "bool"
                }
            ],
            "name": "ApprovalForAll",
            "type": "event"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_owner",
                    "type": "address"
                }
            ],
            "name": "balanceOf",
            "outputs": [
                {
                    "name": "_balance",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "ownerOf",
            "outputs": [
                {
                    "name": "_owner",
                    "type": "address"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "exists",
            "outputs": [
                {
                    "name": "_exists",
                    "type": "bool"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_to",
                    "type": "address"
                },
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "approve",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "getApproved",
            "outputs": [
                {
                    "name": "_operator",
                    "type": "address"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_operator",
                    "type": "address"
                },
                {
                    "name": "_approved",
                    "type": "bool"
                }
            ],
            "name": "setApprovalForAll",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_owner",
                    "type": "address"
                },
                {
                    "name": "_operator",
                    "type": "address"
                }
            ],
            "name": "isApprovedForAll",
            "outputs": [
                {
                    "name": "",
                    "type": "bool"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_from",
                    "type": "address"
                },
                {
                    "name": "_to",
                    "type": "address"
                },
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "transferFrom",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_from",
                    "type": "address"
                },
                {
                    "name": "_to",
                    "type": "address"
                },
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "safeTransferFrom",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_from",
                    "type": "address"
                },
                {
                    "name": "_to",
                    "type": "address"
                },
                {
                    "name": "_tokenId",
                    "type": "uint256"
                },
                {
                    "name": "_data",
                    "type": "bytes"
                }
            ],
            "name": "safeTransferFrom",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        }
    ],
    "bytecode": "0x",
    "deployedBytecode": "0x",
    "sourceMap": "",
    "deployedSourceMap": "",
    "source": "pragma solidity ^0.4.18;\n\n/**\n * @title ERC721 Non-Fungible Token Standard basic interface\n * @dev see https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md\n */\ncontract ERC721Basic {\n  event Transfer(address indexed _from, address indexed _to, uint256 _tokenId);\n  event Approval(address indexed _owner, address indexed _approved, uint256 _tokenId);\n  event ApprovalForAll(address indexed _owner, address indexed _operator, bool _approved);  \n\n  function balanceOf(address _owner) public view returns (uint256 _balance);\n  function ownerOf(uint256 _tokenId) public view returns (address _owner);\n  function exists(uint256 _tokenId) public view returns (bool _exists);\n  \n  function approve(address _to, uint256 _tokenId) public;\n  function getApproved(uint256 _tokenId) public view returns (address _operator);\n  \n  function setApprovalForAll(address _operator, bool _approved) public;\n  function isApprovedForAll(address _owner, address _operator) public view returns (bool);\n\n  function transferFrom(address _from, address _to, uint256 _tokenId) public;\n  function safeTransferFrom(address _from, address _to, uint256 _tokenId) public;  \n  function safeTransferFrom(address _from, address _to, uint256 _tokenId, bytes _data) public;\n}\n",
    "sourcePath": "zeppelin-solidity/contracts/token/ERC721/ERC721Basic.sol",
    "ast": {
        "absolutePath": "zeppelin-solidity/contracts/token/ERC721/ERC721Basic.sol",
        "exportedSymbols": {
            "ERC721Basic": [
                10824
            ]
        },
        "id": 10825,
        "nodeType": "SourceUnit",
        "nodes": [
            {
                "id": 10719,
                "literals": [
                    "solidity",
                    "^",
                    "0.4",
                    ".18"
                ],
                "nodeType": "PragmaDirective",
                "src": "0:24:43"
            },
            {
                "baseContracts": [],
                "contractDependencies": [],
                "contractKind": "contract",
                "documentation": "@title ERC721 Non-Fungible Token Standard basic interface\n@dev see https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md",
                "fullyImplemented": false,
                "id": 10824,
                "linearizedBaseContracts": [
                    10824
                ],
                "name": "ERC721Basic",
                "nodeType": "ContractDefinition",
                "nodes": [
                    {
                        "anonymous": false,
                        "documentation": null,
                        "id": 10727,
                        "name": "Transfer",
                        "nodeType": "EventDefinition",
                        "parameters": {
                            "id": 10726,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10721,
                                    "indexed": true,
                                    "name": "_from",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10727,
                                    "src": "208:21:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10720,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "208:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10723,
                                    "indexed": true,
                                    "name": "_to",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10727,
                                    "src": "231:19:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10722,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "231:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10725,
                                    "indexed": false,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10727,
                                    "src": "252:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10724,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "252:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "207:62:43"
                        },
                        "src": "193:77:43"
                    },
                    {
                        "anonymous": false,
                        "documentation": null,
                        "id": 10735,
                        "name": "Approval",
                        "nodeType": "EventDefinition",
                        "parameters": {
                            "id": 10734,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10729,
                                    "indexed": true,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10735,
                                    "src": "288:22:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10728,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "288:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10731,
                                    "indexed": true,
                                    "name": "_approved",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10735,
                                    "src": "312:25:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10730,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "312:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10733,
                                    "indexed": false,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10735,
                                    "src": "339:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10732,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "339:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "287:69:43"
                        },
                        "src": "273:84:43"
                    },
                    {
                        "anonymous": false,
                        "documentation": null,
                        "id": 10743,
                        "name": "ApprovalForAll",
                        "nodeType": "EventDefinition",
                        "parameters": {
                            "id": 10742,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10737,
                                    "indexed": true,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10743,
                                    "src": "381:22:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10736,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "381:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10739,
                                    "indexed": true,
                                    "name": "_operator",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10743,
                                    "src": "405:25:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10738,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "405:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10741,
                                    "indexed": false,
                                    "name": "_approved",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10743,
                                    "src": "432:14:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_bool",
                                        "typeString": "bool"
                                    },
                                    "typeName": {
                                        "id": 10740,
                                        "name": "bool",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "432:4:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_bool",
                                            "typeString": "bool"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "380:67:43"
                        },
                        "src": "360:88:43"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10750,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "balanceOf",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10746,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10745,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10750,
                                    "src": "473:14:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10744,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "473:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "472:16:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10749,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10748,
                                    "name": "_balance",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10750,
                                    "src": "510:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10747,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "510:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "509:18:43"
                        },
                        "scope": 10824,
                        "src": "454:74:43",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10757,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "ownerOf",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10753,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10752,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10757,
                                    "src": "548:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10751,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "548:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "547:18:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10756,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10755,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10757,
                                    "src": "587:14:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10754,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "587:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "586:16:43"
                        },
                        "scope": 10824,
                        "src": "531:72:43",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10764,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "exists",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10760,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10759,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10764,
                                    "src": "622:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10758,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "622:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "621:18:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10763,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10762,
                                    "name": "_exists",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10764,
                                    "src": "661:12:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_bool",
                                        "typeString": "bool"
                                    },
                                    "typeName": {
                                        "id": 10761,
                                        "name": "bool",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "661:4:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_bool",
                                            "typeString": "bool"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "660:14:43"
                        },
                        "scope": 10824,
                        "src": "606:69:43",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10771,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": false,
                        "modifiers": [],
                        "name": "approve",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10769,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10766,
                                    "name": "_to",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10771,
                                    "src": "698:11:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10765,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "698:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10768,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10771,
                                    "src": "711:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10767,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "711:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "697:31:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10770,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "735:0:43"
                        },
                        "scope": 10824,
                        "src": "681:55:43",
                        "stateMutability": "nonpayable",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10778,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "getApproved",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10774,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10773,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10778,
                                    "src": "760:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10772,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "760:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "759:18:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10777,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10776,
                                    "name": "_operator",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10778,
                                    "src": "799:17:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10775,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "799:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "798:19:43"
                        },
                        "scope": 10824,
                        "src": "739:79:43",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10785,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": false,
                        "modifiers": [],
                        "name": "setApprovalForAll",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10783,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10780,
                                    "name": "_operator",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10785,
                                    "src": "851:17:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10779,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "851:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10782,
                                    "name": "_approved",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10785,
                                    "src": "870:14:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_bool",
                                        "typeString": "bool"
                                    },
                                    "typeName": {
                                        "id": 10781,
                                        "name": "bool",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "870:4:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_bool",
                                            "typeString": "bool"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "850:35:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10784,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "892:0:43"
                        },
                        "scope": 10824,
                        "src": "824:69:43",
                        "stateMutability": "nonpayable",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10794,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "isApprovedForAll",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10790,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10787,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10794,
                                    "src": "922:14:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10786,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "922:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10789,
                                    "name": "_operator",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10794,
                                    "src": "938:17:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10788,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "938:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "921:35:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10793,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10792,
                                    "name": "",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10794,
                                    "src": "978:4:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_bool",
                                        "typeString": "bool"
                                    },
                                    "typeName": {
                                        "id": 10791,
                                        "name": "bool",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "978:4:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_bool",
                                            "typeString": "bool"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "977:6:43"
                        },
                        "scope": 10824,
                        "src": "896:88:43",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10803,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": false,
                        "modifiers": [],
                        "name": "transferFrom",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10801,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10796,
                                    "name": "_from",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10803,
                                    "src": "1010:13:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10795,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1010:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10798,
                                    "name": "_to",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10803,
                                    "src": "1025:11:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10797,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1025:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10800,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10803,
                                    "src": "1038:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10799,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1038:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "1009:46:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10802,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "1062:0:43"
                        },
                        "scope": 10824,
                        "src": "988:75:43",
                        "stateMutability": "nonpayable",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10812,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": false,
                        "modifiers": [],
                        "name": "safeTransferFrom",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10810,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10805,
                                    "name": "_from",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10812,
                                    "src": "1092:13:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10804,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1092:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10807,
                                    "name": "_to",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10812,
                                    "src": "1107:11:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10806,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1107:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10809,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10812,
                                    "src": "1120:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10808,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1120:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "1091:46:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10811,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "1144:0:43"
                        },
                        "scope": 10824,
                        "src": "1066:79:43",
                        "stateMutability": "nonpayable",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10823,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": false,
                        "modifiers": [],
                        "name": "safeTransferFrom",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10821,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10814,
                                    "name": "_from",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10823,
                                    "src": "1176:13:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10813,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1176:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10816,
                                    "name": "_to",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10823,
                                    "src": "1191:11:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10815,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1191:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10818,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10823,
                                    "src": "1204:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10817,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1204:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10820,
                                    "name": "_data",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10823,
                                    "src": "1222:11:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_bytes_memory_ptr",
                                        "typeString": "bytes"
                                    },
                                    "typeName": {
                                        "id": 10819,
                                        "name": "bytes",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1222:5:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_bytes_storage_ptr",
                                            "typeString": "bytes"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "1175:59:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10822,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "1241:0:43"
                        },
                        "scope": 10824,
                        "src": "1150:92:43",
                        "stateMutability": "nonpayable",
                        "superFunction": null,
                        "visibility": "public"
                    }
                ],
                "scope": 10825,
                "src": "168:1076:43"
            }
        ],
        "src": "0:1245:43"
    },
    "legacyAST": {
        "absolutePath": "zeppelin-solidity/contracts/token/ERC721/ERC721Basic.sol",
        "exportedSymbols": {
            "ERC721Basic": [
                10824
            ]
        },
        "id": 10825,
        "nodeType": "SourceUnit",
        "nodes": [
            {
                "id": 10719,
                "literals": [
                    "solidity",
                    "^",
                    "0.4",
                    ".18"
                ],
                "nodeType": "PragmaDirective",
                "src": "0:24:43"
            },
            {
                "baseContracts": [],
                "contractDependencies": [],
                "contractKind": "contract",
                "documentation": "@title ERC721 Non-Fungible Token Standard basic interface\n@dev see https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md",
                "fullyImplemented": false,
                "id": 10824,
                "linearizedBaseContracts": [
                    10824
                ],
                "name": "ERC721Basic",
                "nodeType": "ContractDefinition",
                "nodes": [
                    {
                        "anonymous": false,
                        "documentation": null,
                        "id": 10727,
                        "name": "Transfer",
                        "nodeType": "EventDefinition",
                        "parameters": {
                            "id": 10726,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10721,
                                    "indexed": true,
                                    "name": "_from",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10727,
                                    "src": "208:21:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10720,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "208:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10723,
                                    "indexed": true,
                                    "name": "_to",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10727,
                                    "src": "231:19:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10722,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "231:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10725,
                                    "indexed": false,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10727,
                                    "src": "252:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10724,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "252:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "207:62:43"
                        },
                        "src": "193:77:43"
                    },
                    {
                        "anonymous": false,
                        "documentation": null,
                        "id": 10735,
                        "name": "Approval",
                        "nodeType": "EventDefinition",
                        "parameters": {
                            "id": 10734,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10729,
                                    "indexed": true,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10735,
                                    "src": "288:22:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10728,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "288:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10731,
                                    "indexed": true,
                                    "name": "_approved",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10735,
                                    "src": "312:25:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10730,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "312:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10733,
                                    "indexed": false,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10735,
                                    "src": "339:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10732,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "339:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "287:69:43"
                        },
                        "src": "273:84:43"
                    },
                    {
                        "anonymous": false,
                        "documentation": null,
                        "id": 10743,
                        "name": "ApprovalForAll",
                        "nodeType": "EventDefinition",
                        "parameters": {
                            "id": 10742,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10737,
                                    "indexed": true,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10743,
                                    "src": "381:22:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10736,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "381:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10739,
                                    "indexed": true,
                                    "name": "_operator",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10743,
                                    "src": "405:25:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10738,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "405:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10741,
                                    "indexed": false,
                                    "name": "_approved",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10743,
                                    "src": "432:14:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_bool",
                                        "typeString": "bool"
                                    },
                                    "typeName": {
                                        "id": 10740,
                                        "name": "bool",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "432:4:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_bool",
                                            "typeString": "bool"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "380:67:43"
                        },
                        "src": "360:88:43"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10750,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "balanceOf",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10746,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10745,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10750,
                                    "src": "473:14:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10744,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "473:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "472:16:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10749,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10748,
                                    "name": "_balance",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10750,
                                    "src": "510:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10747,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "510:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "509:18:43"
                        },
                        "scope": 10824,
                        "src": "454:74:43",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10757,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "ownerOf",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10753,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10752,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10757,
                                    "src": "548:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10751,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "548:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "547:18:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10756,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10755,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10757,
                                    "src": "587:14:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10754,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "587:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "586:16:43"
                        },
                        "scope": 10824,
                        "src": "531:72:43",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10764,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "exists",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10760,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10759,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10764,
                                    "src": "622:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10758,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "622:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "621:18:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10763,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10762,
                                    "name": "_exists",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10764,
                                    "src": "661:12:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_bool",
                                        "typeString": "bool"
                                    },
                                    "typeName": {
                                        "id": 10761,
                                        "name": "bool",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "661:4:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_bool",
                                            "typeString": "bool"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "660:14:43"
                        },
                        "scope": 10824,
                        "src": "606:69:43",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10771,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": false,
                        "modifiers": [],
                        "name": "approve",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10769,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10766,
                                    "name": "_to",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10771,
                                    "src": "698:11:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10765,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "698:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10768,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10771,
                                    "src": "711:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10767,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "711:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "697:31:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10770,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "735:0:43"
                        },
                        "scope": 10824,
                        "src": "681:55:43",
                        "stateMutability": "nonpayable",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10778,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "getApproved",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10774,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10773,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10778,
                                    "src": "760:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10772,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "760:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "759:18:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10777,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10776,
                                    "name": "_operator",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10778,
                                    "src": "799:17:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10775,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "799:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "798:19:43"
                        },
                        "scope": 10824,
                        "src": "739:79:43",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10785,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": false,
                        "modifiers": [],
                        "name": "setApprovalForAll",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10783,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10780,
                                    "name": "_operator",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10785,
                                    "src": "851:17:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10779,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "851:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10782,
                                    "name": "_approved",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10785,
                                    "src": "870:14:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_bool",
                                        "typeString": "bool"
                                    },
                                    "typeName": {
                                        "id": 10781,
                                        "name": "bool",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "870:4:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_bool",
                                            "typeString": "bool"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "850:35:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10784,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "892:0:43"
                        },
                        "scope": 10824,
                        "src": "824:69:43",
                        "stateMutability": "nonpayable",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10794,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "isApprovedForAll",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10790,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10787,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10794,
                                    "src": "922:14:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10786,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "922:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10789,
                                    "name": "_operator",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10794,
                                    "src": "938:17:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10788,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "938:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "921:35:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10793,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10792,
                                    "name": "",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10794,
                                    "src": "978:4:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_bool",
                                        "typeString": "bool"
                                    },
                                    "typeName": {
                                        "id": 10791,
                                        "name": "bool",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "978:4:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_bool",
                                            "typeString": "bool"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "977:6:43"
                        },
                        "scope": 10824,
                        "src": "896:88:43",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10803,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": false,
                        "modifiers": [],
                        "name": "transferFrom",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10801,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10796,
                                    "name": "_from",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10803,
                                    "src": "1010:13:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10795,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1010:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10798,
                                    "name": "_to",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10803,
                                    "src": "1025:11:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10797,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1025:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10800,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10803,
                                    "src": "1038:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10799,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1038:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "1009:46:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10802,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "1062:0:43"
                        },
                        "scope": 10824,
                        "src": "988:75:43",
                        "stateMutability": "nonpayable",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10812,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": false,
                        "modifiers": [],
                        "name": "safeTransferFrom",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10810,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10805,
                                    "name": "_from",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10812,
                                    "src": "1092:13:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10804,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1092:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10807,
                                    "name": "_to",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10812,
                                    "src": "1107:11:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10806,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1107:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10809,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10812,
                                    "src": "1120:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10808,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1120:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "1091:46:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10811,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "1144:0:43"
                        },
                        "scope": 10824,
                        "src": "1066:79:43",
                        "stateMutability": "nonpayable",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10823,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": false,
                        "modifiers": [],
                        "name": "safeTransferFrom",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10821,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10814,
                                    "name": "_from",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10823,
                                    "src": "1176:13:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10813,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1176:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10816,
                                    "name": "_to",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10823,
                                    "src": "1191:11:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10815,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1191:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10818,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10823,
                                    "src": "1204:16:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10817,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1204:7:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10820,
                                    "name": "_data",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10823,
                                    "src": "1222:11:43",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_bytes_memory_ptr",
                                        "typeString": "bytes"
                                    },
                                    "typeName": {
                                        "id": 10819,
                                        "name": "bytes",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "1222:5:43",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_bytes_storage_ptr",
                                            "typeString": "bytes"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "1175:59:43"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10822,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "1241:0:43"
                        },
                        "scope": 10824,
                        "src": "1150:92:43",
                        "stateMutability": "nonpayable",
                        "superFunction": null,
                        "visibility": "public"
                    }
                ],
                "scope": 10825,
                "src": "168:1076:43"
            }
        ],
        "src": "0:1245:43"
    },
    "compiler": {
        "name": "solc",
        "version": "0.4.24+commit.e67f0147.Emscripten.clang"
    },
    "networks": {},
    "schemaVersion": "2.0.0",
    "updatedAt": "2018-08-12T00:50:40.743Z"
};
