// @dharmaprotocol/contracts/dist/artifacts/ts/ERC721.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ERC721 = {
    "contractName": "ERC721",
    "abi": [
        {
            "constant": true,
            "inputs": [],
            "name": "name",
            "outputs": [
                {
                    "name": "_name",
                    "type": "string"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "getApproved",
            "outputs": [
                {
                    "name": "_operator",
                    "type": "address"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_to",
                    "type": "address"
                },
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "approve",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "totalSupply",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_from",
                    "type": "address"
                },
                {
                    "name": "_to",
                    "type": "address"
                },
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "transferFrom",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_owner",
                    "type": "address"
                },
                {
                    "name": "_index",
                    "type": "uint256"
                }
            ],
            "name": "tokenOfOwnerByIndex",
            "outputs": [
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_from",
                    "type": "address"
                },
                {
                    "name": "_to",
                    "type": "address"
                },
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "safeTransferFrom",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "exists",
            "outputs": [
                {
                    "name": "_exists",
                    "type": "bool"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_index",
                    "type": "uint256"
                }
            ],
            "name": "tokenByIndex",
            "outputs": [
                {
                    "name": "",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "ownerOf",
            "outputs": [
                {
                    "name": "_owner",
                    "type": "address"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_owner",
                    "type": "address"
                }
            ],
            "name": "balanceOf",
            "outputs": [
                {
                    "name": "_balance",
                    "type": "uint256"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [],
            "name": "symbol",
            "outputs": [
                {
                    "name": "_symbol",
                    "type": "string"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_operator",
                    "type": "address"
                },
                {
                    "name": "_approved",
                    "type": "bool"
                }
            ],
            "name": "setApprovalForAll",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": false,
            "inputs": [
                {
                    "name": "_from",
                    "type": "address"
                },
                {
                    "name": "_to",
                    "type": "address"
                },
                {
                    "name": "_tokenId",
                    "type": "uint256"
                },
                {
                    "name": "_data",
                    "type": "bytes"
                }
            ],
            "name": "safeTransferFrom",
            "outputs": [],
            "payable": false,
            "stateMutability": "nonpayable",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "tokenURI",
            "outputs": [
                {
                    "name": "",
                    "type": "string"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "constant": true,
            "inputs": [
                {
                    "name": "_owner",
                    "type": "address"
                },
                {
                    "name": "_operator",
                    "type": "address"
                }
            ],
            "name": "isApprovedForAll",
            "outputs": [
                {
                    "name": "",
                    "type": "bool"
                }
            ],
            "payable": false,
            "stateMutability": "view",
            "type": "function"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": true,
                    "name": "_from",
                    "type": "address"
                },
                {
                    "indexed": true,
                    "name": "_to",
                    "type": "address"
                },
                {
                    "indexed": false,
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "Transfer",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": true,
                    "name": "_owner",
                    "type": "address"
                },
                {
                    "indexed": true,
                    "name": "_approved",
                    "type": "address"
                },
                {
                    "indexed": false,
                    "name": "_tokenId",
                    "type": "uint256"
                }
            ],
            "name": "Approval",
            "type": "event"
        },
        {
            "anonymous": false,
            "inputs": [
                {
                    "indexed": true,
                    "name": "_owner",
                    "type": "address"
                },
                {
                    "indexed": true,
                    "name": "_operator",
                    "type": "address"
                },
                {
                    "indexed": false,
                    "name": "_approved",
                    "type": "bool"
                }
            ],
            "name": "ApprovalForAll",
            "type": "event"
        }
    ],
    "bytecode": "0x",
    "deployedBytecode": "0x",
    "sourceMap": "",
    "deployedSourceMap": "",
    "source": "pragma solidity ^0.4.18;\n\nimport \"./ERC721Basic.sol\";\n\n/**\n * @title ERC-721 Non-Fungible Token Standard, optional enumeration extension\n * @dev See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md\n */\ncontract ERC721Enumerable is ERC721Basic {\n  function totalSupply() public view returns (uint256);\n  function tokenOfOwnerByIndex(address _owner, uint256 _index) public view returns (uint256 _tokenId);\n  function tokenByIndex(uint256 _index) public view returns (uint256);\n}\n\n/**\n * @title ERC-721 Non-Fungible Token Standard, optional metadata extension\n * @dev See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md\n */\ncontract ERC721Metadata is ERC721Basic {\n  function name() public view returns (string _name);\n  function symbol() public view returns (string _symbol);\n  function tokenURI(uint256 _tokenId) public view returns (string);\n}\n\n/**\n * @title ERC-721 Non-Fungible Token Standard, full implementation interface\n * @dev See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md\n */\ncontract ERC721 is ERC721Basic, ERC721Enumerable, ERC721Metadata {\n}\n",
    "sourcePath": "zeppelin-solidity/contracts/token/ERC721/ERC721.sol",
    "ast": {
        "absolutePath": "zeppelin-solidity/contracts/token/ERC721/ERC721.sol",
        "exportedSymbols": {
            "ERC721": [
                10717
            ],
            "ERC721Enumerable": [
                10690
            ],
            "ERC721Metadata": [
                10710
            ]
        },
        "id": 10718,
        "nodeType": "SourceUnit",
        "nodes": [
            {
                "id": 10665,
                "literals": [
                    "solidity",
                    "^",
                    "0.4",
                    ".18"
                ],
                "nodeType": "PragmaDirective",
                "src": "0:24:42"
            },
            {
                "absolutePath": "zeppelin-solidity/contracts/token/ERC721/ERC721Basic.sol",
                "file": "./ERC721Basic.sol",
                "id": 10666,
                "nodeType": "ImportDirective",
                "scope": 10718,
                "sourceUnit": 10825,
                "src": "26:27:42",
                "symbolAliases": [],
                "unitAlias": ""
            },
            {
                "baseContracts": [
                    {
                        "arguments": null,
                        "baseName": {
                            "contractScope": null,
                            "id": 10667,
                            "name": "ERC721Basic",
                            "nodeType": "UserDefinedTypeName",
                            "referencedDeclaration": 10824,
                            "src": "243:11:42",
                            "typeDescriptions": {
                                "typeIdentifier": "t_contract$_ERC721Basic_$10824",
                                "typeString": "contract ERC721Basic"
                            }
                        },
                        "id": 10668,
                        "nodeType": "InheritanceSpecifier",
                        "src": "243:11:42"
                    }
                ],
                "contractDependencies": [
                    10824
                ],
                "contractKind": "contract",
                "documentation": "@title ERC-721 Non-Fungible Token Standard, optional enumeration extension\n@dev See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md",
                "fullyImplemented": false,
                "id": 10690,
                "linearizedBaseContracts": [
                    10690,
                    10824
                ],
                "name": "ERC721Enumerable",
                "nodeType": "ContractDefinition",
                "nodes": [
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10673,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "totalSupply",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10669,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "279:2:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10672,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10671,
                                    "name": "",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10673,
                                    "src": "303:7:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10670,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "303:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "302:9:42"
                        },
                        "scope": 10690,
                        "src": "259:53:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10682,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "tokenOfOwnerByIndex",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10678,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10675,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10682,
                                    "src": "344:14:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10674,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "344:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10677,
                                    "name": "_index",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10682,
                                    "src": "360:14:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10676,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "360:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "343:32:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10681,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10680,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10682,
                                    "src": "397:16:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10679,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "397:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "396:18:42"
                        },
                        "scope": 10690,
                        "src": "315:100:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10689,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "tokenByIndex",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10685,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10684,
                                    "name": "_index",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10689,
                                    "src": "440:14:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10683,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "440:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "439:16:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10688,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10687,
                                    "name": "",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10689,
                                    "src": "477:7:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10686,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "477:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "476:9:42"
                        },
                        "scope": 10690,
                        "src": "418:68:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    }
                ],
                "scope": 10718,
                "src": "214:274:42"
            },
            {
                "baseContracts": [
                    {
                        "arguments": null,
                        "baseName": {
                            "contractScope": null,
                            "id": 10691,
                            "name": "ERC721Basic",
                            "nodeType": "UserDefinedTypeName",
                            "referencedDeclaration": 10824,
                            "src": "673:11:42",
                            "typeDescriptions": {
                                "typeIdentifier": "t_contract$_ERC721Basic_$10824",
                                "typeString": "contract ERC721Basic"
                            }
                        },
                        "id": 10692,
                        "nodeType": "InheritanceSpecifier",
                        "src": "673:11:42"
                    }
                ],
                "contractDependencies": [
                    10824
                ],
                "contractKind": "contract",
                "documentation": "@title ERC-721 Non-Fungible Token Standard, optional metadata extension\n@dev See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md",
                "fullyImplemented": false,
                "id": 10710,
                "linearizedBaseContracts": [
                    10710,
                    10824
                ],
                "name": "ERC721Metadata",
                "nodeType": "ContractDefinition",
                "nodes": [
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10697,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "name",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10693,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "702:2:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10696,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10695,
                                    "name": "_name",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10697,
                                    "src": "726:12:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_string_memory_ptr",
                                        "typeString": "string"
                                    },
                                    "typeName": {
                                        "id": 10694,
                                        "name": "string",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "726:6:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_string_storage_ptr",
                                            "typeString": "string"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "725:14:42"
                        },
                        "scope": 10710,
                        "src": "689:51:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10702,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "symbol",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10698,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "758:2:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10701,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10700,
                                    "name": "_symbol",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10702,
                                    "src": "782:14:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_string_memory_ptr",
                                        "typeString": "string"
                                    },
                                    "typeName": {
                                        "id": 10699,
                                        "name": "string",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "782:6:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_string_storage_ptr",
                                            "typeString": "string"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "781:16:42"
                        },
                        "scope": 10710,
                        "src": "743:55:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10709,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "tokenURI",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10705,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10704,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10709,
                                    "src": "819:16:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10703,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "819:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "818:18:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10708,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10707,
                                    "name": "",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10709,
                                    "src": "858:6:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_string_memory_ptr",
                                        "typeString": "string"
                                    },
                                    "typeName": {
                                        "id": 10706,
                                        "name": "string",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "858:6:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_string_storage_ptr",
                                            "typeString": "string"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "857:8:42"
                        },
                        "scope": 10710,
                        "src": "801:65:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    }
                ],
                "scope": 10718,
                "src": "646:222:42"
            },
            {
                "baseContracts": [
                    {
                        "arguments": null,
                        "baseName": {
                            "contractScope": null,
                            "id": 10711,
                            "name": "ERC721Basic",
                            "nodeType": "UserDefinedTypeName",
                            "referencedDeclaration": 10824,
                            "src": "1047:11:42",
                            "typeDescriptions": {
                                "typeIdentifier": "t_contract$_ERC721Basic_$10824",
                                "typeString": "contract ERC721Basic"
                            }
                        },
                        "id": 10712,
                        "nodeType": "InheritanceSpecifier",
                        "src": "1047:11:42"
                    },
                    {
                        "arguments": null,
                        "baseName": {
                            "contractScope": null,
                            "id": 10713,
                            "name": "ERC721Enumerable",
                            "nodeType": "UserDefinedTypeName",
                            "referencedDeclaration": 10690,
                            "src": "1060:16:42",
                            "typeDescriptions": {
                                "typeIdentifier": "t_contract$_ERC721Enumerable_$10690",
                                "typeString": "contract ERC721Enumerable"
                            }
                        },
                        "id": 10714,
                        "nodeType": "InheritanceSpecifier",
                        "src": "1060:16:42"
                    },
                    {
                        "arguments": null,
                        "baseName": {
                            "contractScope": null,
                            "id": 10715,
                            "name": "ERC721Metadata",
                            "nodeType": "UserDefinedTypeName",
                            "referencedDeclaration": 10710,
                            "src": "1078:14:42",
                            "typeDescriptions": {
                                "typeIdentifier": "t_contract$_ERC721Metadata_$10710",
                                "typeString": "contract ERC721Metadata"
                            }
                        },
                        "id": 10716,
                        "nodeType": "InheritanceSpecifier",
                        "src": "1078:14:42"
                    }
                ],
                "contractDependencies": [
                    10690,
                    10710,
                    10824
                ],
                "contractKind": "contract",
                "documentation": "@title ERC-721 Non-Fungible Token Standard, full implementation interface\n@dev See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md",
                "fullyImplemented": false,
                "id": 10717,
                "linearizedBaseContracts": [
                    10717,
                    10710,
                    10690,
                    10824
                ],
                "name": "ERC721",
                "nodeType": "ContractDefinition",
                "nodes": [],
                "scope": 10718,
                "src": "1028:68:42"
            }
        ],
        "src": "0:1097:42"
    },
    "legacyAST": {
        "absolutePath": "zeppelin-solidity/contracts/token/ERC721/ERC721.sol",
        "exportedSymbols": {
            "ERC721": [
                10717
            ],
            "ERC721Enumerable": [
                10690
            ],
            "ERC721Metadata": [
                10710
            ]
        },
        "id": 10718,
        "nodeType": "SourceUnit",
        "nodes": [
            {
                "id": 10665,
                "literals": [
                    "solidity",
                    "^",
                    "0.4",
                    ".18"
                ],
                "nodeType": "PragmaDirective",
                "src": "0:24:42"
            },
            {
                "absolutePath": "zeppelin-solidity/contracts/token/ERC721/ERC721Basic.sol",
                "file": "./ERC721Basic.sol",
                "id": 10666,
                "nodeType": "ImportDirective",
                "scope": 10718,
                "sourceUnit": 10825,
                "src": "26:27:42",
                "symbolAliases": [],
                "unitAlias": ""
            },
            {
                "baseContracts": [
                    {
                        "arguments": null,
                        "baseName": {
                            "contractScope": null,
                            "id": 10667,
                            "name": "ERC721Basic",
                            "nodeType": "UserDefinedTypeName",
                            "referencedDeclaration": 10824,
                            "src": "243:11:42",
                            "typeDescriptions": {
                                "typeIdentifier": "t_contract$_ERC721Basic_$10824",
                                "typeString": "contract ERC721Basic"
                            }
                        },
                        "id": 10668,
                        "nodeType": "InheritanceSpecifier",
                        "src": "243:11:42"
                    }
                ],
                "contractDependencies": [
                    10824
                ],
                "contractKind": "contract",
                "documentation": "@title ERC-721 Non-Fungible Token Standard, optional enumeration extension\n@dev See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md",
                "fullyImplemented": false,
                "id": 10690,
                "linearizedBaseContracts": [
                    10690,
                    10824
                ],
                "name": "ERC721Enumerable",
                "nodeType": "ContractDefinition",
                "nodes": [
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10673,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "totalSupply",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10669,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "279:2:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10672,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10671,
                                    "name": "",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10673,
                                    "src": "303:7:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10670,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "303:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "302:9:42"
                        },
                        "scope": 10690,
                        "src": "259:53:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10682,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "tokenOfOwnerByIndex",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10678,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10675,
                                    "name": "_owner",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10682,
                                    "src": "344:14:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_address",
                                        "typeString": "address"
                                    },
                                    "typeName": {
                                        "id": 10674,
                                        "name": "address",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "344:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_address",
                                            "typeString": "address"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                },
                                {
                                    "constant": false,
                                    "id": 10677,
                                    "name": "_index",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10682,
                                    "src": "360:14:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10676,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "360:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "343:32:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10681,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10680,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10682,
                                    "src": "397:16:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10679,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "397:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "396:18:42"
                        },
                        "scope": 10690,
                        "src": "315:100:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10689,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "tokenByIndex",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10685,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10684,
                                    "name": "_index",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10689,
                                    "src": "440:14:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10683,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "440:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "439:16:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10688,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10687,
                                    "name": "",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10689,
                                    "src": "477:7:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10686,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "477:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "476:9:42"
                        },
                        "scope": 10690,
                        "src": "418:68:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    }
                ],
                "scope": 10718,
                "src": "214:274:42"
            },
            {
                "baseContracts": [
                    {
                        "arguments": null,
                        "baseName": {
                            "contractScope": null,
                            "id": 10691,
                            "name": "ERC721Basic",
                            "nodeType": "UserDefinedTypeName",
                            "referencedDeclaration": 10824,
                            "src": "673:11:42",
                            "typeDescriptions": {
                                "typeIdentifier": "t_contract$_ERC721Basic_$10824",
                                "typeString": "contract ERC721Basic"
                            }
                        },
                        "id": 10692,
                        "nodeType": "InheritanceSpecifier",
                        "src": "673:11:42"
                    }
                ],
                "contractDependencies": [
                    10824
                ],
                "contractKind": "contract",
                "documentation": "@title ERC-721 Non-Fungible Token Standard, optional metadata extension\n@dev See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md",
                "fullyImplemented": false,
                "id": 10710,
                "linearizedBaseContracts": [
                    10710,
                    10824
                ],
                "name": "ERC721Metadata",
                "nodeType": "ContractDefinition",
                "nodes": [
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10697,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "name",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10693,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "702:2:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10696,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10695,
                                    "name": "_name",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10697,
                                    "src": "726:12:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_string_memory_ptr",
                                        "typeString": "string"
                                    },
                                    "typeName": {
                                        "id": 10694,
                                        "name": "string",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "726:6:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_string_storage_ptr",
                                            "typeString": "string"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "725:14:42"
                        },
                        "scope": 10710,
                        "src": "689:51:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10702,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "symbol",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10698,
                            "nodeType": "ParameterList",
                            "parameters": [],
                            "src": "758:2:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10701,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10700,
                                    "name": "_symbol",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10702,
                                    "src": "782:14:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_string_memory_ptr",
                                        "typeString": "string"
                                    },
                                    "typeName": {
                                        "id": 10699,
                                        "name": "string",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "782:6:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_string_storage_ptr",
                                            "typeString": "string"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "781:16:42"
                        },
                        "scope": 10710,
                        "src": "743:55:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    },
                    {
                        "body": null,
                        "documentation": null,
                        "id": 10709,
                        "implemented": false,
                        "isConstructor": false,
                        "isDeclaredConst": true,
                        "modifiers": [],
                        "name": "tokenURI",
                        "nodeType": "FunctionDefinition",
                        "parameters": {
                            "id": 10705,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10704,
                                    "name": "_tokenId",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10709,
                                    "src": "819:16:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_uint256",
                                        "typeString": "uint256"
                                    },
                                    "typeName": {
                                        "id": 10703,
                                        "name": "uint256",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "819:7:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_uint256",
                                            "typeString": "uint256"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "818:18:42"
                        },
                        "payable": false,
                        "returnParameters": {
                            "id": 10708,
                            "nodeType": "ParameterList",
                            "parameters": [
                                {
                                    "constant": false,
                                    "id": 10707,
                                    "name": "",
                                    "nodeType": "VariableDeclaration",
                                    "scope": 10709,
                                    "src": "858:6:42",
                                    "stateVariable": false,
                                    "storageLocation": "default",
                                    "typeDescriptions": {
                                        "typeIdentifier": "t_string_memory_ptr",
                                        "typeString": "string"
                                    },
                                    "typeName": {
                                        "id": 10706,
                                        "name": "string",
                                        "nodeType": "ElementaryTypeName",
                                        "src": "858:6:42",
                                        "typeDescriptions": {
                                            "typeIdentifier": "t_string_storage_ptr",
                                            "typeString": "string"
                                        }
                                    },
                                    "value": null,
                                    "visibility": "internal"
                                }
                            ],
                            "src": "857:8:42"
                        },
                        "scope": 10710,
                        "src": "801:65:42",
                        "stateMutability": "view",
                        "superFunction": null,
                        "visibility": "public"
                    }
                ],
                "scope": 10718,
                "src": "646:222:42"
            },
            {
                "baseContracts": [
                    {
                        "arguments": null,
                        "baseName": {
                            "contractScope": null,
                            "id": 10711,
                            "name": "ERC721Basic",
                            "nodeType": "UserDefinedTypeName",
                            "referencedDeclaration": 10824,
                            "src": "1047:11:42",
                            "typeDescriptions": {
                                "typeIdentifier": "t_contract$_ERC721Basic_$10824",
                                "typeString": "contract ERC721Basic"
                            }
                        },
                        "id": 10712,
                        "nodeType": "InheritanceSpecifier",
                        "src": "1047:11:42"
                    },
                    {
                        "arguments": null,
                        "baseName": {
                            "contractScope": null,
                            "id": 10713,
                            "name": "ERC721Enumerable",
                            "nodeType": "UserDefinedTypeName",
                            "referencedDeclaration": 10690,
                            "src": "1060:16:42",
                            "typeDescriptions": {
                                "typeIdentifier": "t_contract$_ERC721Enumerable_$10690",
                                "typeString": "contract ERC721Enumerable"
                            }
                        },
                        "id": 10714,
                        "nodeType": "InheritanceSpecifier",
                        "src": "1060:16:42"
                    },
                    {
                        "arguments": null,
                        "baseName": {
                            "contractScope": null,
                            "id": 10715,
                            "name": "ERC721Metadata",
                            "nodeType": "UserDefinedTypeName",
                            "referencedDeclaration": 10710,
                            "src": "1078:14:42",
                            "typeDescriptions": {
                                "typeIdentifier": "t_contract$_ERC721Metadata_$10710",
                                "typeString": "contract ERC721Metadata"
                            }
                        },
                        "id": 10716,
                        "nodeType": "InheritanceSpecifier",
                        "src": "1078:14:42"
                    }
                ],
                "contractDependencies": [
                    10690,
                    10710,
                    10824
                ],
                "contractKind": "contract",
                "documentation": "@title ERC-721 Non-Fungible Token Standard, full implementation interface\n@dev See https://github.com/ethereum/EIPs/blob/master/EIPS/eip-721.md",
                "fullyImplemented": false,
                "id": 10717,
                "linearizedBaseContracts": [
                    10717,
                    10710,
                    10690,
                    10824
                ],
                "name": "ERC721",
                "nodeType": "ContractDefinition",
                "nodes": [],
                "scope": 10718,
                "src": "1028:68:42"
            }
        ],
        "src": "0:1097:42"
    },
    "compiler": {
        "name": "solc",
        "version": "0.4.24+commit.e67f0147.Emscripten.clang"
    },
    "networks": {},
    "schemaVersion": "2.0.0",
    "updatedAt": "2018-08-12T00:50:40.735Z"
};
